connect/cli/plugins/project/extension/validations.py

```diff
@@ -234,64 +269,173 @@ def validate_schedulables(config, project_dir, context):
+    for _, value in inspect.getmembers(extension_class):
+        if (
+                inspect.isfunction(value)
+                and inspect.getsource(value).strip().startswith('@router.')
+        ):
+            has_router_function = True
+            break
+
+    if not has_router_function:
+        messages.append(
+            ValidationItem(
+                'ERROR',
+                'The Web app extension class must contain at least one router '
+                'implementation function wrapped in *@router.your_method("/your_path")*.',
+                extension_class_file,
+            ),
+        )
+        return ValidationResult(messages, True)
+
+    if 'ui' not in context['descriptor']:
+        messages.append(
+            ValidationItem(
+                'ERROR',
+                'The extension descriptor *extension.json* must contain information '
+                'about static files. Please use *ui* keyword, to define an item '
+                'use *label* for name and *url* to specify absolute path to file within '
+                'static root folder. For more information, look at example: '
+                'https://github.com/cloudblue/eaas-e2e-ma-mock/blob/master/e2e/extension.json.',
+                extension_class_file,
+            ),
+        )
+        return ValidationResult(messages, True)
+
+    ui_items = deque()
+    missed_files = []
+    for _, value in context['descriptor']['ui'].items():
+        ui_items.append(value)
+
+    while ui_items:
+        ui_item = ui_items.pop()
+        try:
+            url = ui_item['url']
+        except KeyError:
             messages.append(
                 ValidationItem(
                     'ERROR',
-                    f'Invalid *secure* attribute for variable *{variable["name"]}*: '
-                    f'must be a boolean not *{type(variable["secure"])}*.',
-                    **get_code_context(extension_class, '@variables'),
+                    'The extension descriptor *extension.json* contains incorrect '
+                    f'ui item *{ui_item.get("label")}*, url is not presented.',
+                    extension_class_file,
                 ),
             )
+            return ValidationResult(messages, True)
+
+        path = os.path.join(os.path.dirname(extension_class_file), url.strip('/'))
+        if not os.path.exists(path):
+            missed_files.append(url)
+
+        for child in ui_item.get('children', []):
+            ui_items.append(child)
+
+    if missed_files:
+        messages.append(
+            ValidationItem(
+                'ERROR',
+                'The extension descriptor *extension.json* contains missing '
+                f'static files: {", ".join(missed_files)}.',
+                extension_class_file,
+            ),
+        )
+        return ValidationResult(messages, True)
+
+    return ValidationResult(messages, False, context)
+
+
+def validate_anvil_extension(config, project_dir, context):
```

Comment:
I think that here you could validate that it is used the @anvil_key_variable and as you said in the comment validate that all the variables are specified properly in the decorator.

Reply:
in decorator you point from where to extract it, not the value itself (which is optional)